### ISSUE TITLE
NTI: No warning for incorrect parameters when using es6 extends and inferred template types

### ISSUE THREAD

**Issue by WearyMonkey** [google/closure-compiler#2232]
```js
/** @template T */
class Root {
  /** @param {T} t */
  constructor(t) {}
}

/** @extends Root<number> */
class Child extends Root { }

class Baa {
  /**
   * @template T
   * @param {function(new: Root<T>, T)} p1
   * @param {T} p2
   */
  static foo(p1, p2) { }
}

// invalid parameter types, T will resolve to both string and number
Baa.foo(Child, 'string');
```

```
java -jar closure-compiler-v20161201.jar -O ADVANCED -W VERBOSE --js source.js --new_type_inf

No errors
```

```
java -jar closure-compiler-v20161201.jar -O ADVANCED -W VERBOSE --js source.js

source.js:20: WARNING - actual parameter 1 of Baa.foo does not match formal parameter
found   : function (new:Child, ...?): undefined
required: function (new:Root<(number|string)>, (number|string)): ?
Baa.foo(Child, 'string');
```

Notes:
Without `extends Root` the error is caught.
Without the `T` constructor parameter in `@param {function(new: Root<T>, T)} p1` the error is caught.
Without the wrapping `class Baa`, the error is not caught by OTI either.

**Comment by dimvar** [comment]
In the following, NTI warns

    /** @template T */
    class Root {
      /** @param {T} t */
      constructor(t) {}
    }
    /** @extends Root<number> */
    class Child extends Root { }
    /**
     * @template T
     * @param {function(new: Root<T>)} p1
     * @param {T} p2
     */
    function foo(p1, p2) { }
    foo(Child, 'string');

The difference from your example is that `p1` takes only one argument. Because of the rules of function subtyping, `Child` in your example is a valid subtype of a function that takes two arguments. But as an unfortunate side effect of that, the `T` in the first argument of `p1` is never matched to anything, so it is matched to the unknown type, and that masks the number and string inconsistency. But the NTI behavior is the correct one here, even though it results in no warning.

**Comment by Dominator008** [comment]
I think we can be a bit stricter here. In this context, `T` can only instantiate as `number`. FYI, TypeScript gives two errors for this case:
```ts
class Root<T> {
  constructor(t: T) {}
}

class Child extends Root<number> { }

interface RootConstructable<T> {
  new(t: T): Root<T>;
}

class Baa {
  static foo<T>(p1: RootConstructable<T>, p2: T) { }
}

// ERROR: The type argument for type parameter 'T' cannot
// be inferred from the usage. Consider specifying the type
// arguments explicitly.
// ERROR: Type argument candidate 'number' is not a valid
// type argument because it is not a supertype of candidate
// '"string"'.
Baa.foo(Child, 'string');
```

**Comment by Dominator008** [comment]
@WearyMonkey Actually, the ES5 equivalent of your original code does trigger a nice warning in NTI:
```js
/**
 * @template T
 * @constructor
 * @param {T} t
 */
function Root(t) {}
/**
 * @constructor
 * @extends Root<number>
 */
function Child() {}
/** @constructor */
function Baa() {}
/**
 * @template T
 * @param {function(new: Root<T>, T)} p1
 * @param {T} p2
 */
Baa.foo = function(p1, p2) {}
Baa.foo(Child, 'string');
```
https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%250A%2520*%2520%2540template%2520T%250A%2520*%2520%2540constructor%250A%2520*%2520%2540param%2520%257BT%257D%2520t%250A%2520*%252F%250Afunction%2520Root(t)%2520%257B%257D%250A%252F**%250A%2520*%2520%2540constructor%250A%2520*%2520%2540extends%2520Root%253Cnumber%253E%250A%2520*%252F%250Afunction%2520Child()%2520%257B%257D%250A%252F**%2520%2540constructor%2520*%252F%250Afunction%2520Baa()%2520%257B%257D%250A%252F**%250A%2520*%2520%2540template%2520T%250A%2520*%2520%2540param%2520%257Bfunction(new%253A%2520Root%253CT%253E%252C%2520T)%257D%2520p1%250A%2520*%2520%2540param%2520%257BT%257D%2520p2%250A%2520*%252F%250ABaa.foo%2520%253D%2520function(p1%252C%2520p2)%2520%257B%257D%250ABaa.foo(Child%252C%2520'string')%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

So something is lost in transpilation here.

**Comment by dimvar** [comment]
Nice catch Michael. Turns out that during transpilation, if the subclass does not have a constructor definition, we add a `var_args` of unknown type. (See below.) This is to support per-file transpilation. So, in this case, there is nothing we can do in type checking. It will be handled when we typecheck ES6 natively.

    /**
     @struct
     @constructor
     @param {T} t
     @template T
     */
    var Root = function(t) {};
    
    /**
     @struct
     @constructor
     @extends {Root<number>}
     @param {...?} var_args
     */
    var Child = function(var_args) {};
    
    /** @struct @constructor */ var Baa = function() {};
    
    /**
     @param {function(new:Root<T>,T)} p1
     @param {T} p2
     @template T

     */
    Baa.foo = function(p1, p2) {};
    
    Baa.foo(Child, "string");

The old type checker gives a warning because it incorrectly ignores unknown types when instantiating polymorphic functions. [Example](https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%2520%250A%2520*%2520%2540template%2520T%250A%2520*%2520%2540param%2520%257BT%257D%2520x%250A%2520*%2520%2540param%2520%257BT%257D%2520y%250A%2520*%2520%2540return%2520%257BT%257D%250A%2520*%252F%250Afunction%2520f(x%252C%2520y)%2520%257B%2520return%2520y%253B%2520%257D%250A%250A%252F%252F%2520The%2520return%2520type%2520should%2520be%2520%253F%252C%2520not%2520number.%250Avar%2520%252F**%2520null%2520*%252F%2520n%2520%253D%2520f(1%252C%2520%252F**%2520%2540type%2520%257B%253F%257D%2520*%252F%2520('asdf'))%253B%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1).
